python/dgl/dataloading/graphsaint.py

```diff
@@ -0,0 +1,128 @@
+
+    Parameters
+    ----------
+    mode : str
+        The sampler to use, which can be :attr:`'node'`, :attr:`'edge'`, or :attr:`'walk'`.
+    budget : int or tuple[int]
+        Sampler configuration.
+
+        - For :attr:`'node'` sampler, budget specifies the number of nodes
+          in each sampled subgraph.
+        - For :attr:`'edge'` sampler, budget specifies the number of edges
+          to sample for inducing a subgraph.
+        - For :attr:`'walk'` sampler, budget is a tuple. budget[0] specifies
+          the number of root nodes to generate random walks. budget[1] specifies
+          the length of a random walk.
+
+    prefetch_ndata : list[str], optional
+        The node data to prefetch for the subgraph.
+
+        See :ref:`guide-minibatch-prefetching` for a detailed explanation of prefetching.
+    prefetch_edata : list[str], optional
+        The edge data to prefetch for the subgraph.
+
+        See :ref:`guide-minibatch-prefetching` for a detailed explanation of prefetching.
+    output_device : device, optional
+        The device of the output subgraphs.
+
+    Examples
+    --------
+
+    >>> import torch
+    >>> from dgl.dataloading import SAINTSampler, DataLoader
+    >>> num_iters = 1000
+    >>> sampler = SAINTSampler(mode='node', budget=6000)
+    >>> dataloader = DataLoader(g, torch.arange(num_iters), sampler, num_workers=4)
+    >>> for subg in dataloader:
+    ...     train_on(subg)
+    """
+    def __init__(self, mode, budget, prefetch_ndata=None,
+                 prefetch_edata=None, output_device='cpu'):
+        super().__init__()
+        self.budget = budget
+        if mode == 'node':
+            self.sampler = self.node_sampler
+        elif mode == 'edge':
+            self.sampler = self.edge_sampler
+        elif mode == 'walk':
+            self.sampler = self.walk_sampler
+        else:
+            raise DGLError(f"Expect mode to be 'node', 'edge' or 'walk', got {mode}.")
+
+        self.prefetch_ndata = prefetch_ndata or []
+        self.prefetch_edata = prefetch_edata or []
+        self.output_device = output_device
+
+    def node_sampler(self, g):
+        """Node ID sampler for random node sampler"""
+        # Alternatively, this can be realized by uniformly sampling an edge subset,
+        # and then take the src node of the sampled edges. However, the number of edges
+        # is typically much larger than the number of nodes.
+        prob = g.out_degrees().float().clamp(min=1)
+        return torch.multinomial(prob, num_samples=self.budget, replacement=True)
+
+    def edge_sampler(self, g):
+        """Node ID sampler for random edge sampler"""
+        src, dst = g.edges()
+        in_deg = g.in_degrees().float().clamp(min=1)
+        out_deg = g.out_degrees().float().clamp(min=1)
+        # We can reduce the sample space by half if graphs are always symmetric.
+        prob = 1. / in_deg[dst.long()] + 1. / out_deg[src.long()]
+        prob /= prob.sum()
```

Comment:
Since `SAINTSampler` is likely a quite important sampler, I suggest adding some system efficiency consideration here. How about we adding a caching option (default being `True`) which will cache the probability array the first time. The docstring needs to tell users to turn it off if they want to use the sampler across different graphs.

Reply:
Done